README.md
# PmgElasticsearchBundle

[![Build Status](https://travis-ci.org/AgencyPMG/ElasticsearchBundle.svg?branch=master)](https://travis-ci.org/AgencyPMG/ElasticsearchBundle)


This is an extremely simple bundle to integrate Elasticsearch into Symfony. It
only provides some configuration and the elasticsearch client.

## Installation

Grab the bundle with composer: 

```bash
composer require pmg/elasticsearch-bundle ~1.0
```

And enable it in your `AppKernel`.

```php
<?php
// app/AppKernel.php

class AppKernel extends Kernel
{
    public function registerBundles()
    {
        $bundles = [
            // ...
            new PMG\ElasticsearchBundle\PmgElasticsearchBundle(),
        );

        // ...

        return $bundles;
    }

    // ...
}
```

## Configuration

You'll use the `pmg_elasticsearch` key in your `config.yml` file. The options
are very similar to what is done with with elasticsearch itself. Most of the
options here are `null` and use the default set by `Elasticsearch\Client`.

You can set up multiple connections:

```yaml
pmg_elasticsearch:
    default_client: example
    clients:
        example:
            connection_class: ~
            connection_factory_class: ~
            connection_pool_class: ~
            selector_class: ~
            serializer_class: ~
            sniff_on_start: ~
            hosts:
                - http://localhost:9200
        another: ~
```

Or just a single client:

```yaml
pmg_elasticsearch:
    connection_class: ~
    connection_factory_class: ~
    connection_pool_class: ~
    selector_class: ~
    serializer_class: ~
    sniff_on_start: ~
    hosts:
        - http://localhost:9200
```

## Services

The `pmg_elasticsearch.client` service will always be the default client. Other
clients can be accessed with `pmg_elasticsear.{name}.client`.

```php

use Symfony\Component\DependencyInjection\ContainerInterface;

/** @var $container ContainerInterface */
$client = $container->get('pmg_elasticsearch.client');

$otherClient = $container->get('pmg_elasticsearch.another.client');
```


CHANGELOG.md
# Change Log

## 4.0.0

### Changed

- Require Symfony ~4.0, drop support for Symfony ~2.7 and ~3.0

## 3.0.0

### Changed

- Require elasticsearch/elasticsearch 2.X OR 5.X. Bumped to a new major version
  to avoid surprising anyone who is using the bundle.

### Fixed

n/a

## 2.0.0

### Changed

- Require elasticsearch/elasticsearch 2.X

### Fixed

n/a

## 1.0.3

### Changed

- `ElasticsearchFactory` is no longer final

### Fixed

- Don't ignore the logging parameter in configuration

## 1.0.2

### Changed

- Add a `.gitattributes` file to exclude things from exports (eg. composer)
- Add this changelog

## 1.0.1

### Fixed

- Actually use the `hosts` argument in configuration


test/TestCase.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

namespace PMG\ElasticsearchBundle;

use PHPUnit\Framework\TestCase as PHPUnitTestcase;

abstract class TestCase extends PHPUnitTestCase
{
    // noop
}


test/bootstrap.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

$loader = require __DIR__.'/../vendor/autoload.php';
$loader->addPsr4('PMG\\ElasticsearchBundle\\', __DIR__);


test/BundleTest.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

namespace PMG\ElasticsearchBundle;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Symfony\Component\HttpFoundation\JsonResponse;

/**
 * This just uses a real HttpKernel to test the bundle integration with its
 * default configuration. Basically we just make a request to elasticsearch
 * to make sure our elasticsearch stuff actually works.
 *
 * @group   acceptance
 */
class BundleTest extends WebTestCase
{
    public function testDefaultElasticsearchConfigurationWorksAndCanMakeRequests()
    {
        $client = $this->createClient();
        $client->request('GET', '/');

        $resp = $client->getResponse();

        $this->assertInstanceOf(JsonResponse::class, $resp);
        $body = json_decode($resp->getContent(), true);

        $this->assertArrayHasKey('tagline', $body);
        $this->assertEquals('You Know, for Search', $body['tagline']);
    }

    protected static function getKernelClass()
    {
        return TestKernel::class;
    }
}


src/ElasticsearchFactory.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

namespace PMG\ElasticsearchBundle;

use Psr\Log\LoggerInterface;
use Psr\Log\NullLogger;

/**
 * Create new elasticsearch objects from container configuration and loggers.
 *
 * @since   0.1
 */
class ElasticsearchFactory
{
    private $params;

    private $logger;

    public function __construct(array $params, LoggerInterface $logger=null)
    {
        $this->params = $params;
        $this->logger = $logger ?: new NullLogger();
    }

    public function create()
    {
        $params = $this->params;
        if (!empty($params['logging'])) {
            $params['logObject'] = $this->logger;
            $params['traceObject'] = $this->logger;
        }

        return $this->newClient($params);
    }

    protected function newClient(array $params)
    {
        return new \Elasticsearch\Client($params);
    }
}


src/DependencyInjection/PmgElasticsearchExtension.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

namespace PMG\ElasticsearchBundle\DependencyInjection;

use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\ContainerInterface;
use Symfony\Component\DependencyInjection\Definition;
use Symfony\Component\DependencyInjection\Reference;
use Symfony\Component\HttpKernel\DependencyInjection\ConfigurableExtension;
use Elasticsearch\ClientBuilder;

/**
 * Our DI container extension.
 *
 * @since   0.1
 */
final class PmgElasticsearchExtension extends ConfigurableExtension
{
    /**
     * {@inheritdoc}
     */
    public function loadInternal(array $config, ContainerBuilder $container)
    {
        foreach ($config['clients'] as $clientName => $clientConfig) {
            $this->addClient($container, $clientName, $clientConfig);
        }

        $container->setAlias('pmg_elasticsearch.client', sprintf(
            'pmg_elasticsearch.%s.client',
            $config['default_client'] ?? current(array_keys($config['clients']))
        ))->setPublic(true);
    }

    private function addClient(ContainerBuilder $container, $name, array $config)
    {
        $factory = $container->setDefinition(
            "pmg_elasticsearch.{$name}.client_builder",
            new Definition(ClientBuilder::class)
        );
        $factory->addMethodCall('setSniffOnStart', [!empty($config['sniff_on_start'])]);
        $factory->addMethodCall('setHosts', [$config['hosts']]);

        if ($config['enable_logging']) {
            $factory->addTag('monolog.logger', [
                'channel'   => 'pmg_elasticsearch'
            ]);
            $factory->addMethodCall('setLogger', [
                new Reference('logger', ContainerInterface::NULL_ON_INVALID_REFERENCE),
            ]);
            $factory->addMethodCall('setTracer', [
                new Reference('logger', ContainerInterface::NULL_ON_INVALID_REFERENCE),
            ]);
        }

        if (!empty($config['connection_factory'])) {
            $factory->addMethodCall('setConnectionFactory', [new Reference($config['connection_factory'])]);
        }

        if (!empty($config['connection_pool'])) {
            $factory->addMethodCall('setConnectionPool', [new Reference($config['connection_pool'])]);
        }

        if (!empty($config['serializer'])) {
            $factory->addMethodCall('setSerializer', [new Reference($config['serializer'])]);
        } elseif (!empty($config['serializer_class'])) {
            $factory->addMethodCall('setSerializer', [$config['serializer_class']]);
        }

        $client = $container->setDefinition(
            "pmg_elasticsearch.{$name}.client",
            new Definition(\Elasticsearch\Client::class)
        );
        $client->setFactory([new Reference("pmg_elasticsearch.{$name}.client_builder"), 'build']);
        $client->setPublic(true);
    }
}


src/DependencyInjection/Configuration.php
<?php
/**
 * This file is part of PMG Elasticsearch Bundle
 *
 * @package     PMG\ElasticsearchBundle
 * @license     http://opensource.org/licenses/mit mit
 * @copyright   PMG <https://www.pmg.co>
 */

namespace PMG\ElasticsearchBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\NodeBuilder;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Elasticsearch\Connections\ConnectionFactoryInterface;
use Elasticsearch\ConnectionPool\ConnectionPoolInterface;
use Elasticsearch\ConnectionPool\AbstractConnectionPool;
use Elasticsearch\ConnectionPool\Selectors\SelectorInterface;
use Elasticsearch\Serializers\SerializerInterface;

/**
 * PmgElasticsearchBundle Configuration Structure.
 *
 * @since   0.1
 */
final class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        [$tree, $root] = $this->createTreeBuilder('pmg_elasticsearch');

        $root
            ->beforeNormalization()
            ->ifTrue(function ($config) {
                    // if we have only a single connection, we need to convert it
                    // to our "multiple clients" format.
                    return is_array($config) && !array_key_exists('clients', $config);
                })
                ->then(function (array $config) {
                    $fixed = [
                        'default_client' => $config['default_client'] ?? 'default',
                    ];
                    unset($config['default_client']);

                    $fixed['clients'] = [
                        $fixed['default_client'] => $config,
                    ];

                    return $fixed;
                })
            ->end();

        $children = $root->children()
            ->append($this->createClientNode())
            ->scalarNode('default_client');

        return $tree;
    }

    private function createClientNode()
    {
        [$tree, $node] = $this->createTreeBuilder('clients');

        $clients = $node
            ->requiresAtLeastOneElement()
            ->useAttributeAsKey('name')
            ->prototype('array')
            ->children();

        $this->addClassNode($clients, 'serializer_class', SerializerInterface::class)
            ->info(sprintf('Alternative to `serializer`. The class to use for the serializer, must implement %s', SerializerInterface::class))
        ->end();

        $clients
            ->scalarNode('connection_factory')
                ->info(sprintf('A service identifier that creates an instance of %s', ConnectionFactoryInterface::class))
            ->end()
            ->scalarNode('connection_pool')
                ->info(sprintf('A service identifier that creates an instance of %s', ConnectionPoolInterface::class))
            ->end()
            ->scalarNode('serializer')
                ->info(sprintf('A service identifier that creates an instance of %s', SerializerInterface::class))
            ->end()
            ->booleanNode('sniff_on_start')->end()
            ->booleanNode('enable_logging')->defaultTrue()->end()
            ->arrayNode('hosts')
                ->defaultValue(['http://localhost:9200'])
                ->requiresAtLeastOneElement()
                ->prototype('scalar')
            ->end();

        return $node;
    }

    private function addClassNode(NodeBuilder $node, $name, $interface)
    {
        return $node->scalarNode($name)
            ->validate()
                ->ifTrue(function ($cls) {
                    return $cls && !class_exists($cls);
                })
                ->thenInvalid('Class %s does not exist')
            ->end()
            ->validate()
                ->ifTrue(function ($cls) use ($interface) {
                    return $cls && !is_subclass_of($cls, $interface);
                })
                ->thenInvalid('%s does not implement or subclass '.$interface)
            ->end();
    }

    private function createTreeBuilder(string $rootName) : array
    {
        // compat: symfony < 4.1
        if (method_exists(TreeBuilder::class, 'getRootNode')) {
            $tree = new TreeBuilder($rootName);
            $root = $tree->getRootNode();
        } else {
            $tree = new TreeBuilder();
            $root = $tree->root($rootName);
        }

        return [$tree, $root];
    }
}
